core-backend/common/src/lib.rs

```diff
@@ -197,23 +197,44 @@ pub trait RuntimeCtx<E: Ext> {
     //
     /// `out_ptr` is the location in memory where `buf` should be written to.
     fn write_output(&mut self, out_ptr: u32, buf: &[u8]) -> Result<(), RuntimeCtxError<E::Error>>;
+}
 
-    fn read_memory_as<T: Sized>(&self, ptr: u32) -> Result<T, RuntimeCtxError<E::Error>> {
-        let mut buf = MaybeUninit::<T>::uninit();
-        let mut_slice =
-            unsafe { slice::from_raw_parts_mut(buf.as_mut_ptr() as *mut u8, size_of::<T>()) };
-        self.read_memory_into_buf(ptr, mut_slice)?;
-        Ok(unsafe { buf.assume_init() })
-    }
+/// Writes object in given memory as bytes.
+///
+/// Safety: due to the fact that given object is Sized and we own them in
+/// the context of calling this function, it's safe to take ptr on the
+/// object and represent it as slice. Object will be dropped after
+/// memory.write already finished execution.
+pub fn write_memory_as<T: Sized>(
+    memory: &mut impl Memory,
+    ptr: u32,
+    obj: T,
+) -> Result<(), MemoryError> {
+    let slice =
+        unsafe { slice::from_raw_parts(&obj as *const T as *const u8, mem::size_of::<T>()) };
```

Comment:
Add comment why it's properly aligned and why is it safe to read/write types between native machine and virtual one